oauth2/src/test/java/com/linecorp/armeria/common/auth/oauth2/OAuth2RefreshBeforeTest.java

```diff
@@ -0,0 +1,104 @@
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import com.linecorp.armeria.client.auth.oauth2.AccessTokenRequest;
+import com.linecorp.armeria.client.auth.oauth2.OAuth2AuthorizationGrant;
+import com.linecorp.armeria.client.logging.LoggingClient;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.testing.junit5.server.ServerExtension;
+
+class OAuth2RefreshBeforeTest {
+    private static final AtomicInteger refreshCount = new AtomicInteger();
+
+    @RegisterExtension
+    static final ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) {
+            sb.service("/token", (ctx, req) -> {
+                final int count = refreshCount.getAndIncrement();
+                final GrantedOAuth2AccessToken token =
+                        GrantedOAuth2AccessToken.builder("token" + count)
+                                                .expiresIn(Duration.ofSeconds(20))
+                                                .build();
+                final HttpResponse response = HttpResponse.of(HttpStatus.OK, MediaType.JSON,
+                                                              token.rawResponse());
+                if (count == 0) {
+                    return response;
+                } else {
+                    // Delay the response to check the cached value is used and the refresh is triggered.
+                    return HttpResponse.delayed(response, Duration.ofSeconds(10));
+                }
+            });
+        }
+    };
+
+    @BeforeEach
+    void setUp() {
+        refreshCount.set(0);
+    }
+
+    @Test
+    void shouldRefreshBeforeExpiry() throws InterruptedException {
+        final AccessTokenRequest accessTokenRequest =
+                AccessTokenRequest.ofClientCredentials("client_id", "client_secret");
+        final OAuth2AuthorizationGrant grant =
+                OAuth2AuthorizationGrant.builder(server.webClient(cb -> {
+                                            cb.decorator(LoggingClient.newDecorator());
+                                        }), "/token")
+                                        .refreshBefore(Duration.ofSeconds(10))
+                                        .accessTokenRequest(accessTokenRequest)
+                                        .build();
+
+        assertThat(grant.getAccessToken().toCompletableFuture().join().accessToken())
+                .isEqualTo("token0");
+        // Drain the first request.
+        server.requestContextCaptor().take();
+        // The first token should be used.
+        assertThat(grant.getAccessToken().toCompletableFuture().join().accessToken())
+                .isEqualTo("token0");
+        assertThat(server.requestContextCaptor().isEmpty()).isTrue();
+        Thread.sleep(10000);
```

Comment:
I'd rather not a test consume 20 seconds, the CI is slow as it is 😅

Reply:
I intended to avoid flakiness but it seems long. Let me try to use 5 seconds for refresh.